File: implementations/java-springboot/examples/src/main/java/com/example/accountservice/account/interfaces/task/PayInterestTaskController.java
package com.example.accountservice.account.interfaces.task;

import com.example.accountservice.account.application.command.PayInterestCommand;
import com.example.accountservice.account.application.command.PayInterestService;
import com.example.accountservice.taskqueue.TaskHandler;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.time.LocalDate;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;

/**
 * The Task Queue input adapter (Interface layer) — just as an HTTP Controller delegates an HTTP
 * request to an Application Service, this Task Controller receives a Task Queue message and calls
 * {@link PayInterestService} (scheduling.md "Task Controller — the Interface layer"). It only
 * delegates, with no conditional branching or business rules, and lets an exception propagate as-is
 * so {@code TaskConsumer} can decide on retry/DLQ (errors are not swallowed).
 */
@Component
@RequiredArgsConstructor
public class PayInterestTaskController implements TaskHandler {

    private final PayInterestService payInterestService;
    private final ObjectMapper objectMapper;

    @Override
    public String taskType() {
        return "account.pay-interest";
    }

    @Override
    public void handle(String payload) throws Exception {
        Payload parsed = objectMapper.readValue(payload, Payload.class);
        payInterestService.payInterest(new PayInterestCommand(parsed.date()));
    }

    // The local payload view owned by this Task Controller — its field names must match the JSON
    // produced by infrastructure/scheduling/InterestPaymentScheduler. It does not share the type,
    // contracting only through the JSON schema, to preserve the layer dependency direction
    // (interfaces
    // does not reference infrastructure).
    private record Payload(LocalDate date) {}
}

File: implementations/kotlin-springboot/examples/src/main/kotlin/com/example/accountservice/account/interfaces/task/PayInterestTaskController.kt
package com.example.accountservice.account.interfaces.task

import com.example.accountservice.account.application.command.PayInterestService
import org.springframework.stereotype.Component
import java.time.LocalDate

/**
 * The Interface input adapter for the Task Queue (`account.pay-interest`) (scheduling.md, "Task
 * Controller — the Interface layer"). Just as an HTTP Controller receives an HTTP request and delegates
 * to an Application Service, this component is the entry point that
 * [com.example.accountservice.taskqueue.TaskHandlerRegistry] calls after deserializing an SQS message.
 *
 * It only delegates to the Command, with no logic of its own — no conditional branching or business
 * rules are added here. Exceptions are thrown as-is.
 * [com.example.accountservice.taskqueue.TaskQueueConsumer] catches them and, instead of deleting the
 * message, leaves it to SQS redelivery (at-least-once) — this differs from the HTTP Controller's
 * `@ExceptionHandler` conversion pattern.
 */
@Component
class PayInterestTaskController(
    private val payInterestService: PayInterestService,
) {
    fun payInterest(payDate: LocalDate) {
        payInterestService.payInterest(payDate)
    }
}
